packages/admin/admin/src/common/buttons/feedback/FeedbackButton.tsx

```diff
@@ -0,0 +1,174 @@
+import { ThreeDotSaving } from "@comet/admin-icons";
+import { Button, ButtonClassKey, ButtonProps, ComponentsOverrides } from "@mui/material";
+import { Theme, useThemeProps } from "@mui/material/styles";
+import * as React from "react";
+
+import { createComponentSlot } from "../../../helpers/createComponentSlot";
+import { ThemedComponentBaseProps } from "../../../helpers/ThemedComponentBaseProps";
+import { Tooltip as CometTooltip } from "../../Tooltip";
+
+export type FeedbackButtonClassKey = "idle" | "loading" | "success" | "fail" | "tooltip" | ButtonClassKey;
+
+type OwnerState = { displayState: FeedbackButtonDisplayState };
+
+const Root = createComponentSlot(Button)<FeedbackButtonClassKey, OwnerState>({
+    componentName: "FeedbackButton",
+    slotName: "root",
+    classesResolver(ownerState) {
+        return [
+            ownerState.displayState === "idle" && "idle",
+            ownerState.displayState === "fail" && "fail",
+            ownerState.displayState === "success" && "success",
+            ownerState.displayState === "loading" && "loading",
+        ];
+    },
+})();
+
+const Tooltip = createComponentSlot(CometTooltip)<FeedbackButtonClassKey>({
+    componentName: "FeedbackButton",
+    slotName: "tooltip",
+})();
+
+export interface FeedbackButtonProps
+    extends ThemedComponentBaseProps<{
+            root: typeof Button;
+            tooltip: typeof CometTooltip;
+        }>,
+        ButtonProps {
+    loading?: boolean;
+    hasErrors?: boolean;
+    loadingIcon?: React.ReactNode;
+    startIcon?: React.ReactNode;
+    endIcon?: React.ReactNode;
+    tooltipSuccessMessage?: string;
+    tooltipErrorMessage?: string;
+}
+
+type FeedbackButtonDisplayState = "idle" | "loading" | "success" | "fail";
+
+export function FeedbackButton(inProps: FeedbackButtonProps) {
+    const {
+        loading,
+        hasErrors,
+        children,
+        variant = "contained",
+        color = "primary",
+        classes,
+        disabled,
+        loadingIcon = <ThreeDotSaving />,
+        startIcon,
+        endIcon,
+        tooltipSuccessMessage,
+        tooltipErrorMessage,
+        slotProps,
+
+        ...restProps
+    } = useThemeProps({
+        props: inProps,
+        name: "CometAdminFeedbackButton",
+    });
+
+    const [displayState, setDisplayState] = React.useState<FeedbackButtonDisplayState>("idle");
+
+    const ownerState: OwnerState = {
+        displayState: displayState,
+    };
+
+    const resolveIconForDisplayState = (displayState: FeedbackButtonDisplayState) => {
+        if (startIcon && displayState !== "loading") {
+            return startIcon;
+        } else if (endIcon && displayState !== "loading") {
+            return endIcon;
+        }
+        if (displayState === "loading") {
+            return loadingIcon;
+        }
```

Comment:
```suggestion
        if (displayState === "loading") {
            return loadingIcon;
        }
        
        return startIcon ?? endIcon
```

I believe this could be simplified to this.